documentation/book/con-the-upgrade-process.adoc

```diff
@@ -0,0 +1,56 @@
+// This module is included in the following assemblies:
+//
+// assembly-upgrading-kafka-versions.adoc
+
+[id='con-the-upgrade-process-{context}']
+= The upgrade process
+
+Depending on the difference between the version being upgraded from and to, the upgrade process works in one of two ways:
+
+* A three phase update. This is typically used for upgrading between Kafka versions differing in the major or minor version number, for example 2.0.0 to 2.1.0.
+* A single phase update. This is typically used for upgrading between Kafka versions differing in the patch number, for example 2.0.0 to 2.0.1.
+
+== Three phase update
+
+When the versions of Kafka use a different interbroker protocol version, or log message format version a three phase update is used.
+
+Phase 1::
+The existing cluster is reconfigured so that the brokers will communicate using the old version of the protocol and using the old message format.
+A rolling restart is necessary so that the brokers are using the new image.
+
+Phase 2::
+All the brokers in the cluster are now running the new image.
+While they all _understand_ the new protocol version, they're all _using_ the old version. 
+Similarly they are using the old message format.
+A rolling restart is necessary for them all to start using the new protocol.
+
+Phase 3::
+You can now update clients, one-by-one, to the new version of the Kafka client libraries.
++
+Once all the clients applications are using the new Kafka client libraries the cluster can be reconfigured so that the brokers use the new message format (using `log.message.format.version`).
+A final rolling restart is necessary.
+
+NOTE: It is possible to revert the version change until the point that `log.message.format.version` is updated in step 3.
+Once producers have sent messages in the new message format and those messages have been appended to broker logs it is not possible to revert to the old version because the old brokers won't be able to read the messages in the log.
+
+User action is required as follows:
+
+. Initially to set `log.message.format.version` in the `Kafka.spec.kafka.config` to the message format version is use by the version of Kafka being upgraded _from_. 
+This step is not necessary if `log.message.format.version` already has the correct value.
```

Comment:
Because the `log.message.format.version` could not be in the Kafka YAML (as for the examples we provide), just adding it will start a rolling update (due to config change) which is not related to the Kafka upgrade. I think we should mention that here.

Reply:
Unless the `inter.broker.protocol.version` is set explicitly (in which case the CO won't override the configured version) the CO will handle setting the `inter.broker.protocol.version` itself. I.e. it will set it initially to the old version then to a rolling update, then set it to the new version/remove it. It's then the users turn to upgrade their apps and change the `log.message.format.version` once they've done so. But the fact that you're asking these questions suggests to me that the documentation isn't doing a good enough job of explaining this.